### PR TITLE
Gracefully drain all smartstack namespaces instead of just the first one

### PR DESCRIPTION
**Pull request by Rob-Johnson**

### CODE REVIEW

File: paasta_tools/drain_lib.py [hunk 1]
```diff
@@ -262,25 +263,56 @@ def spool_url(self, task: DrainTask) -> str:
...
+    async def for_each_registration(self, task: DrainTask, func: Callable[..., T]) -> asyncio.Future[T]:
+        futures = [
+            func(url)
+            for url in self.spool_urls(task)
+        ]
+        return await asyncio.gather(*futures)
+
     async def drain(self, task: DrainTask) -> None:
-        return await self.post_spool(task, 'down')
+        await self.for_each_registration(
```
Comment (solarkennedy) [inline paasta_tools/drain_lib.py]: With these functions, they used to "return" the await function thingy, but now they will return None, is that OK?

**Comment by solarkennedy** [comment]
so far so good I think!

File: paasta_tools/drain_lib.py (hunk 1 above)
Comment (Rob-Johnson) [reply paasta_tools/drain_lib.py]: Yeah, it was just returning None anyway (see the type signature)

File: paasta_tools/drain_lib.py [hunk 2]
```diff
@@ -175,112 +163,94 @@ def __init__(
...
-        info = await self.get_spool(task)
-        if info is None or info["state"] == "up":
-            return False
-        else:
-            return True
+        results = await self.for_each_registration(
+            task,
+            get_spool,
+        )
+        return not all([
```
Comment (mattmb) [inline paasta_tools/drain_lib.py]: Is there any chance results could be an empty list? (`all([]) is True`)

File: yelp_package/dockerfiles/itest/httpdrain/main.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
+package main
```
[the remark this answers is not in the document]
Comment (Rob-Johnson) [reply yelp_package/dockerfiles/itest/httpdrain/main.go]: eh, this was mainly driven by not wanting another huge docker image in the itests (the image for this is 6.13MB vs 685MB for the paastapi image, though admittedly there will be shared layers and whatnot that make this not so bad) and not having to write a whole pyramid app for something so simple (maybe I'm imagining this to be a bigger deal than it is).

this is not a hill I'm willing to die on though, so I'll just write the python version if you feel strongly.

**Comment by Rob-Johnson** [comment]
@EvanKrall can you take another look? I've swapped out the golang for a pyramid service instead and addressed your other feedback

File: paasta_tools/drain_lib.py [hunk 4]
```diff
@@ -175,112 +163,94 @@ def __init__(
...
+            res is None or res["state"] == "up"
+            for res in results
+        ])
 
     async def is_safe_to_kill(self, task: DrainTask) -> bool:
-        info = await self.get_spool(task)
-        if info is None or info["state"] == "up":
+        results = await self.for_each_registration(
```
[the remark this answers is not in the document]
Comment (Rob-Johnson) [reply paasta_tools/drain_lib.py]: true. Can I call YAGNI?

[the remark this answers is not in the document]
Comment (EvanKrall) [reply paasta_tools/drain_lib.py]: totally.

**Pull request merged**